File: my-folder/0438-find-all-anagrams-in-a-string/solution.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int s_length = s.length();
        int p_length = p.length();
        if (s_length < p_length) return {};

        vector<int> indices;
        vector<int> check(26, 0);
        vector<int> window(26, 0);

        for(int i = 0; i < p_length; i++) {
            check[p[i] - 'a']++;
            window[s[i] - 'a']++;
        }

        if(check == window) indices.emplace_back(0);

        for(int i = p_length; i < s_length; i++) {
            window[s[i - p_length] - 'a']--;
            window[s[i] - 'a']++;
            if(window == check) indices.emplace_back(i - p_length + 1);
        }

        return indices;
    }
};

File: my-folder/0438-find-all-anagrams-in-a-string/solution.rb
# @param {String} s
# @param {String} p
# @return {Integer[]}
def find_anagrams(s, p)
    s_length = s.length
    p_length = p.length
    return [] if s_length < p_length

    indices = []
    check = Array.new(26, 0)
    window = Array.new(26, 0)

    p = p.bytes.map { _1 - 97 }
    s = s.bytes.map { _1 - 97 }

    p.each { check[_1] += 1 }

    p_length.times { window[s[_1]] += 1 }
    indices << 0 if check == window

    (s_length - p_length).times do |i|
        window[s[i]] -= 1
        window[s[p_length + i]] += 1
        indices << i + 1 if check == window
    end

    return indices
end
